pyproj/_geod.pyx

```diff
@@ -161,46 +184,151 @@ cdef class Geod:
         double lat2,
         int npts,
         bint radians=False,
+        int initial_idx=1,
+        int terminus_idx=1,
+        object out_lons=None,
+        object out_lats=None,
     ):
         """
         given initial and terminus lat/lon, find npts intermediate points.
+        returns lons, lats buffers
+        """
+        res = \
+            self._inv_or_fwd_intermediate(
+                lon1=lon1,
+                lat1=lat1,
+                lon2_or_azi1=lon2,
+                lat2_or_nan=lat2,
+                npts=npts,
+                del_s=0,
+                radians=radians,
+                initial_idx=initial_idx,
+                terminus_idx=terminus_idx,
+                flags = GEOD_INTER_FLAG_AZIS_DISCARD,
+                out_lons=out_lons,
+                out_lats=out_lats,
+                out_azis=None,
+            )
+
+        return res.lons, res.lats
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    def _inv_or_fwd_intermediate(
+        self,
+        double lon1,
+        double lat1,
+        double lon2_or_azi1,
+        double lat2_or_nan,
+        int npts,
+        double del_s,
+        bint radians,
+        int initial_idx,
+        int terminus_idx,
+        int flags,
+        object out_lons,
+        object out_lats,
+        object out_azis,
+    ) -> GeodIntermediateReturn:
+        """
+        .. versionadded:: 3.1.0
+
+        given initial and terminus lat/lon, find npts intermediate points.
+        using given lons, lats buffers
         """
         cdef Py_ssize_t iii
-        cdef double del_s
         cdef double pazi2
         cdef double s12
         cdef double plon2
         cdef double plat2
         cdef geod_geodesicline line
+
+        cdef bint store_az = \
+            out_azis is not None \
+            or (flags & GEOD_INTER_FLAG_AZIS_MASK) == GEOD_INTER_FLAG_AZIS_KEEP
+
         cdef array.array array_template = array.array("d", [])
-        cdef array.array lats = array.clone(array_template, npts, zero=False)
-        cdef array.array lons = array.clone(array_template, npts, zero=False)
-        cdef PyBuffWriteManager lats_buff = PyBuffWriteManager(lats)
-        cdef PyBuffWriteManager lons_buff = PyBuffWriteManager(lons)
+        if (out_lons is None) \
+                or (out_lats is None) \
+                or (out_azis is None and store_az):
+            if npts == 0:
+                raise GeodError(
```

Comment:
If you define these earlier without initializing them:
```
        cdef PyBuffWriteManager lons_buff
        cdef PyBuffWriteManager lats_buff
        cdef PyBuffWriteManager azis_buff
```
Are you able to move this section after your npts check and this mode will work for all scenarios?

Reply:
good point.
I'm understanding that `array.clone` allocates a Python array, Could I call it inside a `with nogil` section ? 
If not, would there be a performance penalty to make another `with nogil` section after this call ?

Reply:
> Could I call it inside a with nogil section

You would have to call `with gil:` when you allocate the arrays.

> would there be a performance penalty to make another with nogil section after this call ?

There could be a minor penalty if you are doing multi-threading. Single threaded applications won't be impacted.

Reply:
> > Could I call it inside a with nogil section
> 
> You would have to call `with gil:` when you allocate the arrays.
> 
> > would there be a performance penalty to make another with nogil section after this call ?
> 
> There could be a minor penalty if you are doing multi-threading. Single threaded applications won't be impacted.

OK, I've implemented your suggestion.
it's hard for me to evaluate if this case (`inv` + `del_s` + no input array) justifies the performance penalty for all the other cases, but I couldn't find a more elegant solution but to write two versions of this function.
I guess I could investigate that further in another PR.

Reply:
see another comment in the conversation above.

Reply:
Sounds good. Thanks :+1: